substrate/primitives/metadata-ir/src/types.rs

```diff
@@ -56,6 +56,22 @@ pub struct RuntimeApiMetadataIR<T: Form = MetaForm> {
 	pub deprecation_info: DeprecationStatusIR<T>,
 }
 
+impl<T: Form> RuntimeApiMetadataIR<T> {
+	/// This returns a version of `Self` keeping only the
+	/// methods for which the provided function returns true.
+	pub fn keeping_methods<F>(self, f: F) -> Self
```

Comment:
But different methods can be implemented at different versions, so my understanding is that I need to lookat the version that was implemented and then filter out all those methods that haev av ersion greater than this?

Reply:
ie if we have:

```rust
decl_runtime_apis! {
	#[api_version(2)]
	pub trait ApiWithMultipleVersions {
		fn stable_one(data: u64);
		#[api_version(3)]
		fn new_one();
		#[api_version(4)]
		fn glory_one();
	}
}

impl_runtime_apis! {
	#[api_version(3)]
	impl self::ApiWithMultipleVersions<Block> for Runtime {
		fn stable_one(_: u64) {}
		fn new_one() {}
	}
}
```

Then I see that we are implementing version 3 of the APIs, and so I filter out methods declared higher than that (above, `glory_one`) since they haven't been implemented

Reply:
```
fn runtime_metadata(impl_version: u32);

runtime_metadata(version_from_the_attribute);
```

What I propose is to move the filtering to the `runtime_metadata` function. We are calling this function any way and then filter the functions. Instead we could filter directly in `runtime_metadata`.

Reply:
Ooh I see, gotcha! I considered that early on but probably opted not to change the interface :) Lemme try this and see what it looks like!

Reply:
Done! This means `sp-metadata-ir` is untouched now which is nice, though I guess means a major bump on sp-proc-macro since the interface changes a little. The code is a bit simpler I think now though which I like!